src/picongpu/include/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp

```diff
@@ -41,118 +41,182 @@ namespace particles
+         */
+        template< typename ParticleType >
+        HDINLINE uint32_t
+        operator()( float_X const kinEnergyDensity, float_X const density, ParticleType & parentIon, float_X randNr )
+        {
+            float_64 const chargeState = attribute::getChargeState(parentIon);
+            /* @TODO replace the float_64 with float_X and make sure the values are scaled to PIConGPU units */
+            constexpr float_64 protonNumber = GetAtomicNumbers< ParticleType >::type::numberOfProtons;
+
+            /* determine number of new free macro electrons
+             * to be created in the ionization routine
+             */
+            uint32_t numNewFreeMacroElectrons = 0u;
+
+            /* only ionize not-fully ionized ions */
+            if( chargeState < protonNumber )
+            {
+                /* Thomas-Fermi calculation step:
+                 * Determines the new average charge state for each ion under
+                 * LTE conditions.
+                 */
+                float_X const ZStar = detailedBalanceThomasFermi(
+                    kinEnergyDensity,
+                    density,
+                    parentIon
+                );
+
+                /* integral part of the average charge state */
+                float_X intZStar;
+                /* fractional part of the average charge state */
+                float_X const fracZStar = math::modf( ZStar, &intZStar );
+
+                /* Determine new charge state.
+                 * We do a Monte-Carlo step to distribute charge states between
+                 * the two "surrounding" integer numbers if ZStar has a non-zero
+                 * fractional part.
+                 */
+                float_X const newChargeState =
+                    intZStar +
+                    float_X( 1.0 ) * ( randNr < fracZStar );
+
+                /* define number of bound macro electrons before ionization */
+                float_X const prevBoundElectrons = parentIon[ boundElectrons_ ];
 
                 /** determine the new number of bound electrons from the TF ionization state
                  * @TODO introduce partial macroparticle ionization / ionization distribution at some point
                  */
-                float_X const newBoundElectrons = protonNumber - chargeState;
+                float_X const newBoundElectrons = protonNumber - newChargeState;
 
-                return newBoundElectrons;
+                /* Only account for ionization: we only increase the charge
+                 * state of an ion if necessary, but ignore recombination of
+                 * electrons as prediced by the implemented detailed balance
+                 * algorithm.
+                 */
+                if( prevBoundElectrons > newBoundElectrons )
+                    numNewFreeMacroElectrons = static_cast< uint32_t >( prevBoundElectrons - newBoundElectrons );
             }
 
-            return float_X(0.);
+            return numNewFreeMacroElectrons;
```

Comment:
After the last change something went wrong again and we get too many electrons. :cry: 

I will look for it